module/zfs/dmu_objset.c

```diff
@@ -1072,6 +1079,58 @@ dmu_objset_snapshot_one(const char *fsname, const char *snapname)
 	return (err);
 }
 
+typedef struct dmu_objset_upgrade_args {
+	objset_t *dua_os;
+	int (*dua_cb)(objset_t *);
+} dmu_objset_upgrade_arg_t;
+
+static void
+dmu_objset_upgrade_trampoline(void *data)
+{
+	dmu_objset_upgrade_arg_t *arg = data;
+	objset_t *os = arg->dua_os;
+
+	cv_signal(&os->os_upgrade_cv);
+
+	mutex_enter(&os->os_upgrade_lock);
+	if (!os->os_upgrade_exit) {
+		mutex_exit(&os->os_upgrade_lock);
+
+		(void) arg->dua_cb(os);
+		mutex_enter(&os->os_upgrade_lock);
+	}
+	os->os_upgrade_thread = NULL;
+	cv_broadcast(&os->os_upgrade_cv);
+	mutex_exit(&os->os_upgrade_lock);
```

Comment:
> I know this problem, and this is why I put cv_broadcast() inside mutex_exit()

I don't see what's protecting the objset from being destroyed after that first `cv_signal()`.

> I have a concern that retrying wouldn't get rid of the error at all. In that case, the system would just keep spawning threads but couldn't move forward.

What's the alternative?